### ISSUE TITLE
Add '--bail' option when running '--build' mode

### ISSUE THREAD

**Issue by MLoughry** [Microsoft/TypeScript#29938]
## Search Terms
project references, bail, cascade, errors

## Suggestion
When building multiple Typescript projects using the `--build` flag, and the projects had dependencies on each other, an error in an upstream project can cause a cascade of errors from downstream projects that obscure the original error.

A possible solution would be a `--bail` option (or however you'd like to name it). When provided in `--build` mode, it would prevent all projects downstream from a project with an error from building. While this may hide downstream errors that are only found when re-running, it will prevent "false-positive" errors from being shown.

## Use Cases
Using `--build` mode on a large web of project references, without the possibility of being flooded with spurious errors due to one upstream typo.

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by yortus** [comment]
I'm actually curious why bailing isn't the *default* behaviour. That is, why does build mode go ahead building downstream projects when it knows that a dependency failed to build? I have also gotten the cascade of spurious errors described above and been surprised by the current behaviour.

**Comment by RyanCavanaugh** [comment]
My recollection was that bailing was indeed the default - @sheetalkamat thoughts?

**Comment by sheetalkamat** [comment]
yeah bailing should be default. If you are seeing differently then that's a bug and repro would help us investigate it. Thanks.

**Comment by jstevans** [comment]
I believe I [have a small repro](https://github.com/jstevans/ts-build-error-cascade) of the issue we're seeing. 

In our repo, a single erroneous import can cause thousands of these errors.

**Comment by sheetalkamat** [comment]
@jstevans I cant repro this. 
Here is the output::
```
c:\temp\ts-build-error-cascade>node node_modules\typescript\lib\tsc.js -v
Version 3.3.3333

c:\temp\ts-build-error-cascade>node node_modules\typescript\lib\tsc.js -b -v
[10:44:35] Projects in this build:
    * packages/indirectImport/tsconfig.json
    * packages/directImport/tsconfig.json
    * packages/reprosTheBug/tsconfig.json
    * tsconfig.json

[10:44:35] Project 'packages/indirectImport/tsconfig.json' is out of date because output file 'out/indirectImport/index.js' does not exist

[10:44:35] Building project 'c:/temp/ts-build-error-cascade/packages/indirectImport/tsconfig.json'...

[10:44:36] Project 'packages/directImport/tsconfig.json' is out of date because output file 'out/directImport/index.js' does not exist

[10:44:36] Building project 'c:/temp/ts-build-error-cascade/packages/directImport/tsconfig.json'...

[10:44:37] Project 'packages/reprosTheBug/tsconfig.json' is out of date because output file 'out/reprosTheBug/badImport.js' does not exist

[10:44:37] Building project 'c:/temp/ts-build-error-cascade/packages/reprosTheBug/tsconfig.json'...

error TS6307: File 'c:/temp/ts-build-error-cascade/packages/directimport/index.ts' is not in project file list. Projects must list all files or use an 'include' pattern.

error TS6307: File 'c:/temp/ts-build-error-cascade/packages/indirectimport/index.ts' is not in project file list. Projects must list all files or use an 'include' pattern.


Found 2 errors.
```

**Comment by jstevans** [comment]
@sheetalkamat That looks like the repro -- our expectation is that the second `TS6307` shouldn't appear, as it's not directly referenced in any file included by `packages/reprosTheBug/tsconfig.json`.

**Comment by sheetalkamat** [comment]
The compiler options and syntax errors are reported together, what is skipped is doing semantic errors check, and declaration emit errors check.

**Comment by jstevans** [comment]
Ah, that's a helpful distinction. Is there an existing way to suppress the warnings from indirect references, to match our desired behavior?

**Comment by ernestostifano** [comment]
We are also experiencing this issue. Just upgraded TypeScript from `v5.3.3` to `v5.8.3` and we passed from 0 errors to thousands of errors. Probably because there is a new error in our core package (imported by all others) and since it is failing, then all other packages ares failing too.

On `v5.3.3` it was automatically bailing, on `v5.8.3` it continues. This is all the information I have right now.

**Comment by ernestostifano** [comment]
> The compiler options and syntax errors are reported together, what is skipped is doing semantic errors check, and declaration emit errors check.

@sheetalkamat I'm not sure if I fully understood this, but in our case I am seen thousands of type errors due to imported types being `any`. This was not happening the previous version where using `tsc --build` never resulted in "output file does not exist" (using `tsc -p` in a random package produced those errors though).

It is also worth noting that we are using `"customConditions": ["development"]` and that export always point to the source file of each package (maybe has something to do with `types` condition being placed before the `development` one in our `package.json` files. Will check).

**Comment by sheetalkamat** [comment]
This was changed in 5.7 . Refer to https://devblogs.microsoft.com/typescript/announcing-typescript-5-6/#allow---build-with-intermediate-errors To keep older behavior use flag `stopOnBuildErrors`
